parachain/integration-test/src/lib.rs

```diff
@@ -0,0 +1,417 @@
+	chain_b_client: Arc<dyn IsmpProvider>,
+	chain_a_sub_client: SubstrateClient<Hyperbridge>,
+	tx_block_height: u64,
+) -> Result<Vec<u8>, anyhow::Error> {
+	let state_machine_update = StateMachineUpdated {
+		state_machine_id: chain_a_sub_client.state_machine_id(),
+		latest_height: tx_block_height + 2,
+	};
+	let event_result_a = chain_a_client
+		.query_ismp_events(tx_block_height - 2, state_machine_update.clone())
+		.await?;
+
+	// ====================== get the GET_REQUEST ===============================
+	let request_events = event_result_a
+		.into_iter()
+		.filter(|event| match event {
+			Event::GetRequest(_) => true,
+			_ => false,
+		})
+		.collect::<Vec<Event>>();
+
+	// ======== process the request offchain ( 1. Make the response ) =============
+	let get_request = request_events.get(0).unwrap();
+
+	let response = {
+		let get = match get_request {
+			Event::GetRequest(get) => get,
+			_ => panic!("Not supported"),
+		};
+		let dest_chain_block_hash = client_b.rpc().block_hash(Some(get.height.into())).await?;
+		let keys = get.keys.iter().map(|key| &key[..]).collect::<Vec<&[u8]>>();
+		let value_proof = client_b.rpc().read_proof(keys, dest_chain_block_hash).await?;
+		let proof = value_proof.proof.into_iter().map(|bytes| bytes.0).collect::<Vec<Vec<u8>>>();
+
+		let proof_of_value = SubstrateStateProof::StateProof(StateMachineProof {
+			hasher: HashAlgorithm::Keccak,
+			storage_proof: proof,
+		});
+		let proof = Proof {
+			height: StateMachineHeight {
+				id: chain_b_client.state_machine_id(),
+				height: get.height,
+			},
+			proof: proof_of_value.encode(),
+		};
+		let response = ResponseMessage {
+			datagram: RequestResponse::Request(vec![Get(get.clone())]),
+			proof,
+			signer: chain_b_client.address(),
+		};
+
+		Message::Response(response)
+	};
+	// =================== send to the source chain ================================
+	let _res = chain_a_client.submit(vec![response]).await?;
+	//==================== after approx 8-9 blocks the request event is emitted ===
+	// =================== fetch the returned value ================================
+	let mut height_to_fetch = tx_block_height + 8;
+	let mut block_hash = client_a.rpc().block_hash(Some(height_to_fetch.into())).await?.unwrap();
+	let mut response_event: Option<GetResponse> = None;
+
+	while height_to_fetch <= (tx_block_height + 10) {
+		let fetched_events = client_a.events().at(block_hash).await?.find_first::<GetResponse>()?;
+		match fetched_events {
+			Some(res_event) => {
+				response_event = Some(res_event.clone());
+				break
+			},
+			None => {
+				height_to_fetch += 1;
+				block_hash =
+					client_a.rpc().block_hash(Some(height_to_fetch.into())).await?.unwrap();
+			},
+		}
+	}
+
+	let encoded_value = match response_event.unwrap().0[0].clone() {
+		Some(value) => value,
+		None => {
+			panic!("Value not found")
+		},
+	};
+
+	Ok(encoded_value)
+}
+
+/// Configure the state machines and relayer
+async fn initial_setup() -> Result<
```

Comment:
Returning only the substrate client is sufficient for all your needs
```suggestion
async fn initial_setup() -> Result<(SubstrateClient<Hyperbridge>, SubstrateClient<Hyperbridge>)>
```

Reply:
That's much cleaner

Reply:
Am refactoring now but others arent duplicate and they are different types in which constructing them you would need the Substrate Config